src/Statement/StatementList.php

```diff
@@ -87,7 +87,13 @@ public function getPropertyIds() {
 	}
 
 	public function addStatement( Statement $statement ) {
-		$this->statements[] = $statement;
+		if ( $statement->getGuid() !== null
+			&& !array_key_exists( $statement->getGuid(), $this->statements )
+		) {
+			$this->statements[$statement->getGuid()] = $statement;
+		} else {
+			$this->statements[] = $statement;
```

Comment:
Uh, having string and integer keys in the same array seems really ugly and indicates that there is a problem in the design of this class. If we want indexing by GUID, we have to enforce all statements in this class to have a GUID.